server/aesdsocket.c

```diff
@@ -130,102 +235,180 @@ int createStreamSocket(const char *portNumberStr)
 {
-
     struct sockaddr peeraddr;
     socklen_t peer_addr_size = sizeof(peeraddr);
     
     int connectedSock = accept(sockfd, &peeraddr, &peer_addr_size);
     if(connectedSock==-1)
     {
-        perror("listen() error");
+        if(errno==EINTR)
+        {
+            //Allow this to keep executing even
+            //on interruptions from interval alarm
+            return -2;
+        }
+
+        perror("accept() error");
         return -1;
     }
 
+    *newListElement = (socket_data_t *)malloc(sizeof(socket_data_t));
+
+    if(*newListElement==NULL)
+    {
+        //If cannot malloc anymore space
+        return -1;
+    }
+
+    //Initialize arguments to be used by thread
+    (*newListElement)->connectedSock = connectedSock;
+    (*newListElement)->peeraddr = peeraddr;
+    (*newListElement)->threadCompleteFlag = false;
+
+    if(pthread_create((&(*newListElement)->threadHandle), NULL,\
+                    recvAndSendAndLog, *newListElement) != 0)
+    {
+        perror("pthread_create() error");
+        free(*newListElement);
+        return -1;
+    }
+
+    return 0;
+}
+
+void* recvAndSendAndLog(void* socket_data_arg)
+{
+    socket_data_t *socket_data = (socket_data_t *) socket_data_arg;  
+
     //Determine IP address of client for logging
     struct sockaddr_in *peeraddr_in;
-    peeraddr_in = (struct sockaddr_in *)(&peeraddr);
+    peeraddr_in = (struct sockaddr_in *)(&(socket_data->peeraddr));
     const char *ipv4Addr = (char *)(&peeraddr_in->sin_addr.s_addr);
 
-    syslog(LOG_INFO, "Accepted connection from %d.%d.%d.%d", ipv4Addr[0], ipv4Addr[1], ipv4Addr[2], ipv4Addr[3]);
+    syslog(LOG_INFO, "Accepted connection from %d.%d.%d.%d",\
+            ipv4Addr[0], ipv4Addr[1], ipv4Addr[2], ipv4Addr[3]);
 
     //Open output file to append to or create if it does not already exist
-    int outputFd = open(OUTPUT_FILEPATH, O_RDWR | O_CREAT | O_APPEND, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP);
+    int outputFd = open(OUTPUT_FILEPATH, O_RDWR | O_CREAT | O_APPEND,\
+                            S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP);
 
     if(outputFd==-1)
     {
-        perror("creat() error");
-        return -1;
+        perror("open() error");
+        socket_data->threadCompleteFlag = true;
+        pthread_exit(socket_data);
     }
 
     //Receive a single byte at a time
     char recvdByte;
-    static int totalByteCnt = 0;
 
     char retByte;
     ssize_t sendRet;
     ssize_t readRet;
+    int lockRet;
+
+    bool mutexLocked=false;
 
     //Append each byte received to the output file
-    while(recv(connectedSock, &recvdByte, 1, 0)!=0)
+    while(recv(socket_data->connectedSock, &recvdByte, 1, 0)!=0)
     {
-        
+        if(!mutexLocked)
+        {
+
+            lockRet = pthread_mutex_lock(&mutex);
+
+            if(lockRet!=0)
+            {
+                perror("pthread_mutex_lock() error");
+                socket_data->threadCompleteFlag = true;
```

Comment:
@ECleverito I don't think you need mutex locking here.  Since threadCompleteFlag starts with state false before the thread starts and you only check for true status set it shouldn't matter whether other threads read while it's being updated.